Centralize the filter lookup in the core actor

Neverlord opened the pull request:
# Problem

The core actor dispatches incoming messages (from local publishers, peers or WebSockets) to subscribers (again: local subscribers, peers or WebSockets). For that, Broker currently sets up data flow pipelines. Each subscriber subscribes to the central merge point (where all the inputs flow together) and then filters unwanted messages.

The problem is that this filtering is done by each subscriber individually. This means we will perform n filter checks, where n is the number of subscribers. Since subscriptions in Broker are prefix-based, this means a filter check isn’t just a single hash map lookup.

# Proposed Solution

At the end of the day, Broker shoves data from one set of buffers (inputs) to another set of buffers (outputs). Those buffers then connect to hubs, stores, peers or WebSocket clients. While data flows come with nice properties and APIs, we barely use any of that. In fact, the pipeline setup in the core actor has become quite involved since there we have added WebSocket peers and hubs to the mix.

The proposed solution operates much closer to the buffers. We introduce a new `handler` concept in the core actor, which is basically just a data sink. It is managing an output buffer and also adds back-pressure. Each handler has a queue that fills up if the buffer reached its capacity (which happens if we push faster than the reader consumes from the buffer). If that queue overflows, we apply the configured strategy (disconnect, drop-oldest or drop-newest). The main change is that we store all handlers in a single filter: a multi map that allows storing multiples handlers for the same topic(s). This means we always do a single, central filter lookup. The result of that lookup is a list of handlers that have a matching subscription. Then we call `offer` on all those handlers. This method may do a few extra checks. For example, a data store handler discards anything that isn’t a `command` message.

With this design, we only ever have exactly one filter lookup. While that filter may still grow with the number of peers, a central filter lookup should still be faster than doing n lookups.

# Benchmarking

Setup: sending 20k messages/s. To start the receiving node:

```
./build/bin/broker-node -t "[$( tr -d '\n' < topics.txt ) '/benchmark/events']" -p 1234 -v -r
```

To start the sending node:

```
./build/bin/broker-throughput --verbose -r 20000 localhost:1234
```

After a while, we peer 20 additional nodes with the receiver, starting each via:

```
./build/bin/broker-node -t "[$( tr -d '\n' < topics.txt )]" --peers='["tcp://localhost:1234"]' -v
```

## Baseline

### No additional peerings

broker-node: ~36% CPU
broker-throughput: ~34% CPU

### 20 additional peerings

broker-node: ~53%
broker-throughput: ~37%

## Branch

broker-node: ~36% CPU
broker-throughput: ~34% CPU

### 20 additional peerings

broker-node: ~47%
broker-throughput: ~36%

# Discussion

Doing a single filter lookup reduces CPU load in the benchmark by about 6% (from 53% to 47%), which is good. However, there is still some linear scaling with the number of peers. The additional peers are not subscribed to the `/benchmark/events` topic, so they don’t receive any data. I can verify from the system diagnostics that the extra peers sit idle well below 1% CPU and don’t receive any data. Hence, they shouldn’t cause additional CPU load.

Long story short, the new structure cuts down unnecessary CPU load due to centralizing the filter lookup (resulting in a ~6% reduction of CPU load for the benchmark). I think the new code is also simpler, which is an added bonus. However, even in the new branch we can see that adding more peers adds run-time cost even if they don’t receive anything. We should look into that next, after wrapping up this PR.

Neverlord committed:
Add new subscription_multimap class

Currently, we have one filter per subscriber in the core actor. Then, we
ask each subscriber individually whether a certain topic matches its
filter.

The new `subscription_multimap` flips this work flow: we register all
subscribers at the topics they are interested in. Then, we make a single
pass over the subscriptions to obtain all matches via the `select`
member function. The use case we have in Broker is to have a single
filter that stores all subscribers and then do a single central lookup.
The runtime cost of `select` depends on the number of subscribed topics,
not on the number of peers. Hence, we should expect better scalability
when switching to `subscription_multimap`.

ckreibich commented on libbroker/broker/internal/subscription_multimap.hh:
```diff
@@ -0,0 +1,85 @@
+    std::sort(result.begin(), result.end());
+    result.erase(std::unique(result.begin(), result.end()), result.end());
+  }
+  std::vector<value_type> select(const key_type& topic) const {
```
Might want an extra line before this for readability.

Neverlord replied:
Done. I've also added more documentation.

ckreibich commented on libbroker/broker/internal/core_actor.cc:
```diff
@@ -1157,18 +1157,10 @@ caf::error core_actor_state::init_new_client(const network_info& addr,
   using impl_t = handler_impl<data_message>;
   auto client_id = endpoint_id::random();
-  auto state = std::make_shared<impl_t>(web_socket_buffer_size(),
+  auto state = std::make_shared<impl_t>(client_id, std::move(type),
```
Should this be in the previous commit?

ckreibich commented on libbroker/broker/internal/core_actor.hh:
```diff
@@ -60,15 +60,18 @@ public:
     ok,
     /// The callback was invoked but resulted in a terminal state.
-    disposed,
+    disconnect,
```
Thank you because I didn't like `disposed` and was going to comment on that. :-)

ckreibich reviewed (changes requested):
Thanks! As usual, sorry for taking forever to review this. My usual Broker disclaimer applies — I think I get this at a high level, and almost certainly won't notice bugs in the details. I am continuing to treat CAF as a black box.

My gut feel is that this is quite a large changeset for a potential 6% improvement, but I do like the more explicit callbacks and overall layout of the code. I'm nervous about new buffering semantics. To be clear: the only real change in backpressure logic is for datastores, right?

My only real change request here is refactoring the commits that start with "fix" into earlier commits — these should ideally go away, unless I'm misunderstanding their purpose? If they need to stay, it'd be great to get more detail on what "fixing" means.

Neverlord commented on libbroker/broker/internal/core_actor.hh:
```diff
@@ -56,6 +62,202 @@ public:
+  public:
+    using super = handler;
+
+    using buffer_producer_ptr = caf::async::spsc_buffer_producer_ptr<T>;
```
Single Producer Single Consumer.

Neverlord committed:
Refactor push and pull callbacks on the observer

Currently, we pass the message that is entering / leaving a peer or
client buffer to the observer. However, this message isn't really useful
to an observer such as Zeek. Furthermore, this API prevents us from
batching calls to the observer. Instead of passing the message to those
callbacks, we now simply pass a `count` argument. This allows us to
"compress" multiple events and also doesn't require us to have the
messages available when calling the observer.

Neverlord committed:
Centralize the filter lookup in the core actor

A recent benchmark revealed that Broker's CPU usage goes up
significantly with the number of peering relationships, even if the new
peers never receive any data.

The root cause for this behavior is the flow-based processing in Broker,
where each new peer results in a new observer that subscribes to the
central flow of messages. For each incoming message, its topic is then
checked against the peer's filter. Hence, each message causes N filter
lookups where N is the number of peers.

Moreover, Broker does not really utilize the feature set of CAF flows.
Ultimately, the core actor simply dispatches data by reading from
buffers filled by local producers or peers and then forwards those
messages by writing to another set of buffers that are connected to
subscribers.

The new `handler` abstraction in the core actor expresses this work flow
more aptly. A handler is simply a combination of a sink and a source. A
handler can be registered at a `subscription_multimap` for each topic it
is interested in. A handler may still ignore matched messages, for
example, a handler for a data store would ignore any message on the
topic that is not of type `command_message`.

Neverlord committed:
Remove obsolete utility classes

After switching from the flow-based processing in the core actor to a
new buffer-oriented approach, we no longer need this set of utilities.

Neverlord commented:
@ckreibich thank you for the feedback! I've split the PR into four logical commits: two prepare steps, the actual change, and one cleanup commit that only removes obsolete files. The unrelated change to the broker-node has been filed separately and I've added additional documentation to the multimap.

Neverlord committed:
Abort output on backpressure instead of closing

By calling `abort` instead of just `dispose` on the output buffer, the
WebSocket client should emit an error frame before closing the
connection to let the clients know that an error occurred.

Neverlord committed:
Add new unit tests for the back-pressure logic

Add three new unit tests for the core actor to verify that the overflow
policy for peers is enforced correctly.

Neverlord committed:
CAF bump for bugfix backport to blocking_consumer

Neverlord committed:
Add additional metrics to the core actor

Neverlord committed:
Signal backpressure_overflow in peer_removed event

Neverlord committed:
Separate message handler and core actor code

This commit has *no functional changes*.

Rename `core_actor::handler` to `message_handler` and move it to its own
header and source files. Also move all dependent classes and enum types.

Removed the `handler_base` template for simpler and cleaner code.

Neverlord committed:
Move the pull buffer to the message handler

Refactoring step that pushes the central pull buffer into the message
handler. Also tweak the interface to allow the caller to stop consuming
pulled items mid-way.

Neverlord committed:
Implement back-pressure for message handlers

Stores and hubs can suspend fast senders via back-pressure. The core
actor now keeps a list of suspended handlers and attempts to auto-resume
suspended handlers whenever a handler received additional demand,
thereby potentially allowing previously blocked data to go through.

Neverlord committed:
Prevent accidental forwarding of messages

When disabling forwarding, Broker will look at the sender ID to know
whether a message comes from the local node or not in order to decide
whether connected peers should receive the message. Sending messages
without sender ID at all bypasses this detection and causes many
replicated messages in full-mesh setups like in the `store.put-unique`
btest.

Neverlord commented:
Too late for 8.2
